### PR TITLE
support for non-correlated subqueries

### LINKED ISSUE #3266: Add support for non-correlated subqueries
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I would like to be able to run this query:

```
CREATE TABLE paintings AS SELECT 'Mona Lisa' as name, 1000000 as listed_price;

SELECT name, listed_price
FROM paintings
WHERE listed_price > (
    SELECT AVG(listed_price)
    FROM paintings
);
```

It currently fails with:

```
Skipping optimizer rule decorrelate_scalar_subquery due to unexpected error: scalar subqueries must have a filter to be correlated at /home/andy/git/apache/arrow-datafusion/datafusion/optimizer/src/decorrelate_scalar_subquery.rs:177
caused by
Error during planning: Could not coerce into Filter! at /home/andy/git/apache/arrow-datafusion/datafusion/expr/src/logical_plan/plan.rs:1127
NotImplemented("Physical plan does not support logical expression (<subquery>)")

```

This is because the existing subquery optimizer rules fail to rewrite the query.

It should be possible to rewrite the query as a join:

```
SELECT name, listed_price
FROM paintings
CROSS JOIN (SELECT AVG(listed_price) AS avg_price FROM paintings) temp
WHERE listed_price > temp.avg_price;
```

**Describe the solution you'd like**
:point_up: 

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by kmitchener**
# Which issue does this PR close?

Closes #3266.

 # Rationale for this change

# What changes are included in this PR?

Handle case where subquery doesn't have a filter.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/decorrelate_scalar_subquery.rs [hunk 1]
```diff
@@ -101,16 +101,17 @@ impl OptimizerRule for DecorrelateScalarSubquery {
...
                     // regular filter, no subquery exists clause here
+                    let optimized_plan = LogicalPlan::Filter(Filter {
+                        predicate: predicate.clone(),
+                        input: Arc::new(optimized_input),
+                    });
                     return Ok(optimized_plan);
                 }
 
-                // iterate through all exists clauses in predicate, turning each into a join
+                // iterate through all subqueries in predicate, turning each into a join
```
Comment (avantgardnerio) [inline datafusion/optimizer/src/decorrelate_scalar_subquery.rs:114]: Thanks for fixing my messy docs and comments everywhere!

File: datafusion/optimizer/src/decorrelate_scalar_subquery.rs [hunk 2]
```diff
@@ -173,20 +191,28 @@ fn optimize_scalar(
         .map_err(|e| context!("Exactly one input is expected. Is this a join?", e))?;
     let aggr = Aggregate::try_from_plan(sub_input)
         .map_err(|e| context!("scalar subqueries must aggregate a value", e))?;
-    let filter = Filter::try_from_plan(&aggr.input).map_err(|e| {
-        context!("scalar subqueries must have a filter to be correlated", e)
-    })?;
+    let filter = Filter::try_from_plan(&aggr.input).ok();
+
+    // if there were filters, we use that logical plan, otherwise the plan from the aggregate
+    let input: &LogicalPlan;
```
Comment (avantgardnerio) [inline datafusion/optimizer/src/decorrelate_scalar_subquery.rs:197]:
```suggestion
    let input = if filter.is_some() {
        input = &filter.unwrap().input;
    } else {
        input = &aggr.input;
    };
```

**Comment by avantgardnerio** [comment]
Thanks for looking at this! It's good to have another person touching this code. I'm happy to help in any way if I can!

File: datafusion/optimizer/src/decorrelate_scalar_subquery.rs (hunk 1 above)
Comment (kmitchener) [reply datafusion/optimizer/src/decorrelate_scalar_subquery.rs:114]: it's very well commented! the comments were extremely helpful

File: datafusion/optimizer/src/decorrelate_scalar_subquery.rs (hunk 2 above)
Comment (kmitchener) [reply datafusion/optimizer/src/decorrelate_scalar_subquery.rs:197]: That won't compile. I googled for how to do this more Rust-like, but didn't find anything more concise that works.

**Comment by kmitchener** [comment]
@andygrove PTAL

**Comment by kmitchener** [comment]
I tested the SQL in #3266 and that works locally. As well as the query in TPCH 15.

File: datafusion/optimizer/src/decorrelate_scalar_subquery.rs (hunk 2 above)
Comment (avantgardnerio) [reply datafusion/optimizer/src/decorrelate_scalar_subquery.rs:197]: Sorry, I was going to pull and compile, but I see you undrafted, so here's what I think will compile and avoid the unwrap() (which we're trying to elliminate for audit purposes even if it should be safe):

```
    let input = match filter {
        Some(filter) => &filter.input,
        None => &aggr.input
    };
    ```

**Comment by avantgardnerio** [comment]
If query 15 works, I'd suggest adding a test here https://github.com/apache/arrow-datafusion/blob/7aed4d697fa24053d515babfd7678855451c6736/datafusion/core/tests/sql/subqueries.rs#L31

File: datafusion/optimizer/src/decorrelate_scalar_subquery.rs [hunk 3]
```diff
@@ -702,4 +727,31 @@ mod tests {
         assert_optimized_plan_eq(&DecorrelateScalarSubquery::new(), &plan, expected);
         Ok(())
     }
+
+    /// Test for non-correlated scalar subquery with no filters
```
Comment (avantgardnerio) [inline datafusion/optimizer/src/decorrelate_scalar_subquery.rs:731]: Should we rename this if it also does uncorrelated subqueries now? Maybe `scalar_subquery_to_join`?

File: datafusion/optimizer/src/decorrelate_scalar_subquery.rs (hunk 2 above)
Comment (kmitchener) [reply datafusion/optimizer/src/decorrelate_scalar_subquery.rs]: Clippy just made me dance, it's all rust-like now :)

File: datafusion/optimizer/src/decorrelate_scalar_subquery.rs [hunk 4]
```diff
@@ -702,4 +727,31 @@ mod tests {
...
+    #[test]
+    fn scalar_subquery_non_correlated_no_filters() -> Result<()> {
```
Comment (avantgardnerio) [inline datafusion/optimizer/src/decorrelate_scalar_subquery.rs:733]: Looks like 1 branch was added, and 1 test. LGTM!

**Review by avantgardnerio — APPROVED** [review]
The unwrap is gone, a test was added for the additional code path... LGTM!

File: datafusion/optimizer/src/decorrelate_scalar_subquery.rs (hunk 3 above)
Comment (kmitchener) [reply datafusion/optimizer/src/decorrelate_scalar_subquery.rs:731]: You mean rename the whole rule? Probably so .. good idea. You want to do that as part of this issue?

Comment (avantgardnerio) [reply datafusion/optimizer/src/decorrelate_scalar_subquery.rs:731]: I don't want to, but we probably should. Maybe we can just doc it for now with a TODO and address it in a future PR?

Comment (kmitchener) [reply datafusion/optimizer/src/decorrelate_scalar_subquery.rs:731]: It touches a bunch of code, but yes, your suggested name is much more clear. I'll do a separate PR for it this week if this gets merged in.

**Review by andygrove — APPROVED** [review]
Thanks @kmitchener. 

Changes look reasonable to me, but I am mostly approving based on the review from @avantgardnerio, who wrote the original code. Thanks for the review @avantgardnerio!

**Pull request merged**